Select with multiple enabled does not consistently fire OnChange

Ahmad-S792 opened the pull request:
#### 066cf30607d538a2f7766f4acf615fb0a9118a4c
<pre>
Select with multiple enabled does not consistently fire OnChange

<a href="https://bugs.webkit.org/show_bug.cgi?id=257565">https://bugs.webkit.org/show_bug.cgi?id=257565</a>

Reviewed by Ryosuke Niwa.

This patch aligns WebKit with Blink / Chromium and Gecko / Firefox.

Partial Merge: <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=164563

This patch adds logic for the case, when the listbox is scrollable to
fire &apos;onChange&apos; event when the selection is dropped outside of &apos;listbox&apos;.

* Source/WebCore/page/AutoscrollController.cpp:
(AutoscrollController::startAutoscrollForSelection): As above
* LayoutTests/fast/forms/select/multiselect-in-listbox-mouse-release-outside.html: Add Test Case
* LayoutTests/fast/forms/select/multiselect-in-listbox-mouse-release-outside-expected.txt: Add Test Case Expectation
* LayoutTests/platform/ios/TestExpectations: &apos;iOS&apos; exception due to lack of &apos;listbox&apos; appearance

Canonical link: <a href="https://commits.webkit.org/264873@main">https://commits.webkit.org/264873@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4bb72ac1fb7a93dbdba7bc1379d0db13a6a2c27b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8690 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8980 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10347 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8690 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8699 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10968 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8951 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11566 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8836 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9852 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7839 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10504 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7152 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7947 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15466 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8268 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8095 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11440 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8577 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6998 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7842 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7858 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2168 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12054 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8322 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/page/AutoscrollController.cpp:
```diff
@@ -75,6 +77,8 @@ void AutoscrollController::startAutoscrollForSelection(RenderObject* renderer)
         return;
     auto* scrollable = RenderBox::findAutoscrollable(renderer);
+    if (!scrollable)
+        scrollable = renderer->isListBox() ? downcast<RenderListBox>(renderer) : nullptr;
```
Why are we leaving AutoscrollController::updateDragAndDrop in tact?
It seems like we should be fixing both functions.

Ahmad-S792 replied:
Dropping does not lead to 'text' / label of options being presented in any Notes or other application, so I think 'onChange' on start would be appropriate. Similarly, if we don't have 'dropTargetNode', which if I understand correctly would mean it need 'node', will just early return as 'nullptr', if we lack, which is the case in this, if we do it in 'updateDragAndDrop' function.

Ahmad-S792 replied:
@rniwa - From Slack 1-1: If I understand function correctly ‘updateDragAndDrop’ needs dropTarget but option text / label does not get dropped and don’t have node so I think it will early return and will not work

rniwa replied:
I see.

Ahmad-S792 committed:
Select with multiple enabled does not consistently fire OnChange

https://bugs.webkit.org/show_bug.cgi?id=257565

Reviewed by Ryosuke Niwa.

This patch aligns WebKit with Blink / Chromium and Gecko / Firefox.

Partial Merge: https://src.chromium.org/viewvc/blink?view=revision&revision=164563

This patch adds logic for the case, when the listbox is scrollable to
fire 'onChange' event when the selection is dropped outside of 'listbox'.

* Source/WebCore/page/AutoscrollController.cpp:
(AutoscrollController::startAutoscrollForSelection): As above
* LayoutTests/fast/forms/select/multiselect-in-listbox-mouse-release-outside.html: Add Test Case
* LayoutTests/fast/forms/select/multiselect-in-listbox-mouse-release-outside-expected.txt: Add Test Case Expectation
* LayoutTests/platform/ios/TestExpectations: 'iOS' exception due to lack of 'listbox' appearance

Canonical link: https://commits.webkit.org/264873@main
